File: src/test/java/com/freya02/botcommands/test/commands/slash/SlashModal.java
package com.freya02.botcommands.test.commands.slash;

import com.freya02.botcommands.api.BContext;
import com.freya02.botcommands.api.commands.application.ApplicationCommand;
import com.freya02.botcommands.api.commands.application.slash.GuildSlashEvent;
import com.freya02.botcommands.api.commands.application.slash.annotations.JDASlashCommand;
import com.freya02.botcommands.api.modals.Modals;
import com.freya02.botcommands.api.modals.annotations.ModalData;
import com.freya02.botcommands.api.modals.annotations.ModalHandler;
import com.freya02.botcommands.api.modals.annotations.ModalInput;
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent;
import net.dv8tion.jda.api.interactions.components.Modal;
import net.dv8tion.jda.api.interactions.components.text.TextInputStyle;

import java.util.concurrent.TimeUnit;

public class SlashModal extends ApplicationCommand {
	private static final String MODAL_HANDLER_NAME = "test";
	private static final String CODE_INPUT_NAME = "code";

	@JDASlashCommand(
			name = "modal",
			description = "Test modal"
	)
	public void onSlashModal(GuildSlashEvent event, Modals modals) {
		final Modal modal = modals.create(MODAL_HANDLER_NAME, "foobar", 42L)
				.setTimeout(10, TimeUnit.SECONDS, () -> System.out.println("bruh"))
				.setTitle("Formatting !")
				.addActionRow(
						modals.createTextInput(CODE_INPUT_NAME, "Java code", TextInputStyle.PARAGRAPH)
								.setRequired(false)
								.build()
				)
				.build();

		event.replyModal(modal).queue();
	}

	@ModalHandler(name = MODAL_HANDLER_NAME)
	public void handle(ModalInteractionEvent event,
	                   @ModalData String test,
	                   @ModalData long number,
	                   @ModalInput(name = CODE_INPUT_NAME) String javaCode,
	                   BContext context) {
		event.reply(("""
						User data: %s
						Context: %s
						Your code:
						%s""").formatted(test + " = " + number, context, javaCode))
				.setEphemeral(true)
				.queue();
	}
}

File: src/test/kotlin/com/freya02/botcommands/test/commands_kt/slash/SlashModal.kt
package com.freya02.botcommands.test.commands_kt.slash

import com.freya02.botcommands.api.annotations.CommandMarker
import com.freya02.botcommands.api.commands.application.ApplicationCommand
import com.freya02.botcommands.api.commands.application.CommandScope
import com.freya02.botcommands.api.commands.application.GuildApplicationCommandManager
import com.freya02.botcommands.api.commands.application.annotations.AppDeclaration
import com.freya02.botcommands.api.commands.application.slash.GuildSlashEvent
import com.freya02.botcommands.api.commands.application.slash.annotations.JDASlashCommand
import com.freya02.botcommands.api.components.Components
import com.freya02.botcommands.api.components.event.ButtonEvent
import com.freya02.botcommands.api.modals.Modals
import com.freya02.botcommands.api.modals.annotations.ModalData
import com.freya02.botcommands.api.modals.annotations.ModalHandler
import com.freya02.botcommands.api.modals.annotations.ModalInput
import com.freya02.botcommands.test.CustomObject
import dev.minn.jda.ktx.interactions.components.row
import dev.minn.jda.ktx.messages.reply_
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent
import net.dv8tion.jda.api.interactions.components.text.TextInputStyle
import java.util.concurrent.TimeUnit

private const val SLASH_MODAL_MODAL_HANDLER = "SlashModal: modalHandler"
private const val SLASH_MODAL_TEXT_INPUT = "SlashModal: textInput"

@CommandMarker
class SlashModal(private val components: Components) : ApplicationCommand() {
    @JDASlashCommand(name = "modal_annotated")
    fun onSlashModal(event: GuildSlashEvent, modals: Modals) {
        val input = modals.createTextInput(SLASH_MODAL_TEXT_INPUT, "Sample text", TextInputStyle.SHORT)
            .build()

        val modal = modals.create("Title", SLASH_MODAL_MODAL_HANDLER, "User data", 420)
            .setTimeout(30, TimeUnit.SECONDS) {
                println("Timeout")
            }
            .addActionRow(input)
            .build()

        event.replyModal(modal).queue()
    }

    @ModalHandler(name = SLASH_MODAL_MODAL_HANDLER)
    fun onModalSubmitted(
        event: ModalInteractionEvent,
        @ModalData dataStr: String,
        @ModalData dataInt: Int,
        @ModalInput(name = SLASH_MODAL_TEXT_INPUT) inputStr: String,
        customObject: CustomObject
    ) {
        event.reply_("""
            Submitted:
            dataStr: $dataStr
            dataInt: $dataInt
            inputStr: $inputStr
            customObject: $customObject
            """.trimIndent(),
            components = listOf(row(components.primaryButton(this::handleButton).build("Test button"))),
            ephemeral = true).queue()
    }

    private fun handleButton(event: ButtonEvent) {
        event.deferEdit().queue()

        println(event.message.interaction?.user?.asMention)
    }

    @AppDeclaration
    fun declare(applicationCommandManager: GuildApplicationCommandManager) {
        applicationCommandManager.slashCommand("modal", scope = CommandScope.GUILD) {
            customOption("modals")

            function = ::onSlashModal
        }
    }
}